### ISSUE TITLE
Provide prepopulation and maintenance of the JDBC connection pool

### ISSUE THREAD

**Issue by frowe** [OpenLiberty/open-liberty#14900]
## Description

As a server administrator, I want to be able to configure a number of connections to a database that will be created and maintained by the server, so I can ensure that applications will not have to wait for connection creation, allowing them to meet specific service level agreements (SLA) for response times.

For this use case, we need to provide the ability to configure a number of connections to be created at JDBC service startup and the ability to maintain these connections to avoid network timeouts and/or inactivity closure.  Creation of the initial connections will need to be spaced such that it doesn't overwhelm the connected resource.  Whether and how the connections are maintained by the server should be configurable.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Aha: Externally raised RFE ([Aha](https://cloud-platform.ideas.aha.io/))
  - https://cloud-platform.ideas.ibm.com/ideas/TWAS-I-229
- UFO: Link to [Upcoming Feature Overview](https://ibm.box.com/v/UFO-Template) document
  - https://ibm.box.com/s/etww1bllwuveqthnw0kra20rayeurooa
- FTS: Link to Feature Test Summary GH Issue
- Beta Blog: Link to Beta Blog Post GH Issue
- GA Blog: Link to GA Blog Post GH Issue

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [ ] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [ ] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an [Upcoming Feature Overview (UFO)](https://ibm.box.com/v/UFO-Template) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.

### **Design Preliminaries**
- [ ] UI requirements identified. (Owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [ ] ID requirements identified. (Owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers))
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [ ] Serviceability Requirements Identified. (Owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [ ] SVT Requirements Identified. (Owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [ ] Performance testing requirements identified. (Owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [ ] POC Design / UFO review requested.
  - Owner adds label `Design Review Request`
- [ ] POC Design / UFO review scheduled.
  - Follow the instructions in POC-Forum repo
- [ ] POC Design / UFO review completed.
- [ ] POC / UFO Review follow-ons completed.
- [ ] POC Design / UFO approval requested.
  - Owner adds label `Design Approval Request`
- [ ] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [ ] No Design requested.
  - Owner adds label `No Design Approval Request`
- [ ] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`
- [ ] Feature / Capability stabilization or discontinuation or N/A
  - Owner adds label `Product Management Approval Request` and notifies [OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management)
  - Approver adds label `Product Management Approved` ([OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management))
  - Note: For stabilized, superseded, and discontinued feature/capability, skip the [Beta](#beta) section of the template (you may delete it).  Otherwise, proceed as normal.

### **FAT Documentation**
- [ ] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) before any implementation work may begin to be delivered (inaccessible/no-ship).  However, a design focused approach should still be applied to features, and developers should think about the feature design prior to writing and delivering any code.  
Besides being prioritized, a feature must also be socialized (or No Design Approved) before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the "Design Approved" or "No Design Approved" label, along with all other tasks outlined in the [GA](#ga) section.

### **Feature Development Begins**
- [ ] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [ ] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).
- [ ] Licenses and Certificates of Originality (COOs) are updated, or N/A

### **Translation** (Complete 1 week before Feature Complete Date)
- [ ] PII updates are merged, or N/A. Note timing with translation shipments.

### **Innovation** (Complete 1 week before Feature Complete Date)
- [ ] Consider whether any aspects of the feature may be patentable. If any identified, disclosures have been submitted.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [ ] Beta fence the functionality
  - `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [ ] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [ ] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete 1.5 weeks before beta eGA)
- [ ] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [beta release blog post](https://openliberty.io/blog/?search=beta&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [ ] Feature implementation and tests completed.
  - [ ] All PRs are merged.
  - [ ] All epic and child issues are closed.
  - [ ] All stop ship issues are completed.
- [ ] Legal: all necessary approvals granted.
- [ ] Translation: All messages translated or sent for translation for upcoming release 
- [ ] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [ ] **APIs/Externals** Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [ ] **Demo** Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Add comment `@OpenLiberty/demo-approvers Demo scheduled for EOI [Iteration Number]` to this issue.
  - Approver adds label `focalApproved:demo`.
- [ ] **FAT** All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.
- [ ] **Globalization** Translation and TVT are complete or N/A. ([OpenLiberty/globalization-approvers](https://github.com/orgs/OpenLiberty/teams/globalization-approvers))
  - Approver adds label `focalApproved:globalization`.

### **Design Approved Features**
- [ ] **Accessibility** Accessibility testing completed or N/A. ([OpenLiberty/accessibility-approvers](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
  - Approver adds label `focalApproved:accessibility`.
- [ ] **ID** Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [ ] **Performance** Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [ ] **Serviceability** Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [ ] **STE** Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [ ] **SVT** System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [ ] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Feature Complete Date)
- [ ] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [release blog post](https://openliberty.io/blog/?search=release&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

### **Post GA**
- [ ] Replace `target:YY00X` label with the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **Standalone Feature Blog Post** A blog post specifically about your feature or N/A.  ([OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **OL Guides** OL Guides assessment is complete or N/A. ([OpenLiberty/guide-assessment](https://github.com/orgs/OpenLiberty/teams/guide-assessment/members))
- [ ] **Dev Experience** Developer Experience & Tools work is complete or N/A. ([OpenLiberty/dev-experience-assessment](https://github.com/orgs/OpenLiberty/teams/dev-experience-assessment/members))
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by frowe** [comment]
UFO socialization issues:
Note, as discussed on several slides, the exact syntax for the `minPoolSizeMaintainedAt` attribute will be determined in a separate Design Issues meeting.  Once agreed up, the UFO will be updated with that info.
Slide 33 bullet 3 - it was noticed that feature should use Liberty standard convention for handling config problems, the onError attribute (fail/warn/ignore)

**Comment by NottyCode** [comment]
As discussed off GitHub I think we should drop the auth alias for maintaining the connection pool minimum and instead just rely on the container auth alias.

**Comment by samwatibm** [comment]
RFE's have moved:

- DevWorks http://www.ibm.com/developerworks/rfe/execute?use_case=viewChangeRequest&CR_ID=144070 maps to:
RFE: https://cloud-platform.ideas.aha.io/ideas/TWAS-I-229


Can't find the equivalent for http://www.ibm.com/developerworks/rfe/execute?use_case=viewChangeRequest&CR_ID=70807 because it's now closed

**Comment by frowe** [comment]
@samwatibm
> Can't find the equivalent for http://www.ibm.com/developerworks/rfe/execute?use_case=viewChangeRequest&CR_ID=70807 because it's now closed
Not sure what you mean by "Can't find equivalent..."

**Comment by samwatibm** [comment]
@frowe External RFEs in DeveloperWorks has migrated to Aha and will eventually go away.   (See the warning on top of your 1st RFE, [144070](https://www.ibm.com/developerworks/rfe/execute?use_case=viewChangeRequest&CR_ID=144070).)  

[70807](www.ibm.com/developerworks/rfe/execute?use_case=viewChangeRequest&CR_ID=70807) is closed so it was never migrated to Aha.   If the info there is important, please capture it somewhere.

**Comment by frowe** [comment]
@malincoln Per your request for status, this issue is not currently in progress, but as we're wrapping up jakarta ee10 work, we'll restart work on it.

**Comment by malincoln** [comment]
@njr-11 as we work through re-prioritizing features under Core Runtime, Alasdair would like an update on features.  I see this one has been paused for a while.  Can you provide the relative importance of this work along with sizing?  Thanks!

**Comment by njr-11** [comment]
> [@njr-11](https://github.com/njr-11) as we work through re-prioritizing features under Core Runtime, Alasdair would like an update on features. I see this one has been paused for a while. Can you provide the relative importance of this work along with sizing? Thanks!

It is still on hold while we work on Jakarta EE 11, and now MORE. There will not be any spare capacity before moving on to Jakarta EE 12, which will overlap the aforementioned work, so it should continue to remain paused.

**Comment by malincoln** [comment]
@njr-11 what's the sizing to complete the work needed for this one?

**Comment by njr-11** [comment]
> [@njr-11](https://github.com/njr-11) what's the sizing to complete the work needed for this one?

3XL. It was barely started before we suspended it.

**Comment by malincoln** [comment]
It's been here for 4.5 years. Need to understand the validity since it's been opened for a while.

**Comment by njr-11** [comment]
This is still a valid requirement that would benefit some customers, but they do have a partial/more primitive workaround of writing application code that requests a bunch of connections upon startup to fill the pool, although it wouldn't be advisable to try to do that to ensure a number of connections at later points when the server may already be doing useful work on connections.

**Comment by malincoln** [comment]
after discussion with Alasdair, Mac, Nathan, Chuck decision made to close Aha idea for now as it would take a great amount of resources and not enough customer interests.
